### DISCUSSION TITLE
How to activate generic typing inference on function intersection type?

### DISCUSSION THREAD

**Discussion by Athelian** [flow-typed/flow-typed#4399] [Q&A]
Hello, sorry if this is a simple question

I am really struggling to figure out how to activate the second version of this set function, see 
[flowtry](https://flow.org/try/#0FAehAIBMFMDMEsB28Au8D2jywE7oLbgA28ARsCgJ4AO04AytCuALzDgecBk4AFL+lIAraAGMUALnAB5YWJQAacNQCGKABYB+KZoDOKHEgDmSgG4qiAV2hSViSgEpWAPhlzxD9pw49eAHgAVKVN0eEhwAB9wREsiImdeL28OQRFxKQCFJOTVDW1wPQNjLOTOcytofM07SmyOJxZXAM8KGjoAMXR0VnAAbwjYLqlSLqJoOwiAX2BRTH1sIfBO7pY+wfQpA2tp0AgambnmXSYAtsgpRmZVwITU+QylXK0dfUNEE3By61t7BwyXcB3cQHRDzRDQADuAEF7D1jihTrRILx1koAETrNFKWAWY4tMBLLog+bwgCqiCoSJ6NwE7kk4EyyjUzwKr2Knws33ANT+DIBQJQxOY4IhyzhTHJlOgyNR4AxXSx2Fx0AcQA)

The definition is taken straight from [here](https://flow.org/try/#0FAehAIBMFMDMEsB28Au8D2jywE7oLbgA28ARsCgJ4AO04AytCuALzDgecBk4AFL+lIAraAGMUALnAB5YWJQAacNQCGKABYB+KZoDOKHEgDmSgG4qiAV2hSViSgEpWAPhlzxD9pw49eAHgAVKVN0eEhwAB9wREsiImdeL28OQRFxKQCFJOTVDW1wPQNjLOTOcytofM07SmyOJxZXAM8KGjoAMXR0VnAAbwjYLqlSLqJoOwiAX2BRTH1sIfBO7pY+wfQpA2tp0AgambnmXSYAtsgpRmZVwITU+QylXK0dfUNEE3By61t7BwyXcB3cQHRDzRDQADuyx6xxQp1okF46yUACJ1iilLALMcWmAll0QfNYQBVRBUBE9G4CdyScCZZRqZ4FV7FT4Wb7gGp-OkAoEoQnMcEQgCC9hhTFJ5OgiOR4DRXQx2Gx0AcQA)

To me I should not be getting an explicit `any` on `newAny`. I also tried activating the generic version using underbar. From Flow docs: 

> If you only want to specify some of the type arguments, you can use _ to let flow infer a type for you:

Please someone let me know if they know how to properly use functions like this.

**Comment by Brianzchen** [comment]
Both links are the same. Not really sure what your issue is. Can you review it a bit then get back to me?

**Reply by Athelian** [reply]
Thanks for the reply.

To me I'm seeing two definitions for `set` in this library.

In this [FlowTry](https://flow.org/try/#0FAehAIBMFMDMEsB28Au8D2jywE7oLbgA28ARsCgJ4AO04AytCuALzDgecBk4AFL+lIAraAGMUALnAB5YWJQAacNQCGKABYB+KZoDOKHEgDmSgG4qiAV2hSViSgEpWAPhlzxD9pw49eAHgAVKVN0eEhwAB9wREsiImdeL28OQRFxKQCFJOTVDW1wPQNjLOTOcytofM07SmyOJxZXAM8KGjoAMXR0VnAAbwjYLqlSLqJoOwiAX2BRTH1sIfBO7pY+wfQpA2tpmbnmXSYewITU+QylXK0dfUNEE3By61t7BwyXcFPxYF3EeYOUAJtSBSRjMVb-H5-JgAVUQVFo4XBTG+YHANUhzEQ0AA7gBBew9f6AhG8dZKABE63JSlgFgOTlRGngunAzPA+BUOAA1tBwioWejQBBlhjojjloSYXCgaSuhSqTS6dAGRAmSy2Rzuby0SzlkA):
Line 14:  define a function `set`

Line 16: define a variable `setTyped`, annotated with type `Set` and equal to the `set` function 
Line 17: define a variable `setUntyped`, the same as above but unannotated

Result:
`setTyped`'s return type is always `any`
`setUntyped`s return type uses generics correctly.

When I invoke `lodash.set` with these typings in a production repo, every return is marked as `any`.

I would like to know how to invoke `lodash.set` using the definition:
`
(<T: void | null>(
        object: T,
        path?: ?string,
        value?: ?any
      ) => T)
`

and not using the definition:
`
((object: Object, path?: ?string, value: any) => Object)
`

**Reply by pascalduez** [reply]
Side note: `Object` does not exist anymore in Flow, and is resolving to `any`, so you'd want to avoid using it, it gives false assumptions.

**Reply by pascalduez** [reply]
Also `<T: void | null>` is quite weird, it's a type constraint and means you'd only allow `void | null` types.

**Reply by Athelian** [reply]
> Also `<T: void | null>` is quite weird, it's a type constraint and means you'd only allow `void | null` types.

To me this is saying "Do not add anything in here. If this is empty, then it will activate Flow's type inference system and we will infer the return type". 

But I cannot get the type inference to activate in the first place with this current definition.
